### PR TITLE
Jetstream by default

### PR DESCRIPTION
**Pull request by tengomucho**
# What does this PR do?

This makes all the changes to allow having the Jetstream Pytorch engine to be the default backend for TGI on TPUs.
This backend is reliable and performant and give the best throughput on TGI.

### CODE REVIEW

**Review by dacorvo — COMMENTED** [review]
I think I got lost in your changes: can you summarize how tests are now supposed to work ?

File: text-generation-inference/tests/test_generator_slot.py [hunk 1]
```diff
@@ -15,6 +18,54 @@ def tokenizer(request):
...
+            " slipped quickly through the glass doors of Victory Mansions, though not quickly enough"
+            " to prevent a swirl of gritty dust from entering along with him.",
+        ],
+        ["This sentence is written in chinese:", "我很感谢你的热情"],
+        ["Some text might contain a lot of emojis like 😃", "😍💪 👉 👀"],
+    ],
+    ids=["spaces", "chinese-utf8", "emojis"],
+)
+def test_decode_streaming_jetstream(tokenizer, input_text, generated_text):
+    if not jetstream_pt_available():
```
[the remark this answers is not in the document]
Comment (tengomucho) [reply text-generation-inference/tests/test_generator_slot.py]: I refactored the test to avoid repetitions.

File: text-generation-inference/tests/test_prefill_truncate.py [hunk 2]
```diff
@@ -1,35 +1,47 @@
...
     batch = Batch(id=0, requests=[request], size=1, max_tokens=sequence_length)
     generations, _ = generator.prefill(batch)
     assert len(generations) == 1
-    assert generations[0].tokens.ids == [291]
-    assert generations[0].tokens.texts == [" and"]
+    assert generations[0].tokens.ids == [266]
+    assert generations[0].tokens.texts == [" the"]
+
+
+def test_prefill_truncate_jetstream():
```
[the remark this answers is not in the document]
Comment (tengomucho) [reply text-generation-inference/tests/test_prefill_truncate.py:37]: The two tests are identical in behaviour, but if jetstream is loaded the other test will fail to run correctly because of incompatibility on the dependencies when using some features of pytorch (i.e.: multiprocessing).
I just have two identical tests, but one is going to be run when jetstream is enabled, the other one will be skipped, and when jetstream is disabled it will be the other way around.

File: text-generation-inference/tests/conftest.py [hunk 3]
```diff
@@ -33,3 +35,14 @@ def quantization_jetstream_int8():
...
     os.environ.update(old_environ)
+
+
+def pytest_runtest_setup(item):
+    marker_names = [marker.name for marker in item.own_markers]
+    jetstream_pt_enabled = jetstream_pt_available()
+    # Skip tests that require torch xla but not jetstream
+    if "torch_xla" in marker_names and "jetstream" not in marker_names:
+        if jetstream_pt_enabled:
+            pytest.skip("Jetstream PyTorch must be disabled")
```
Comment (dacorvo) [inline text-generation-inference/tests/conftest.py:46]: nit: I would find it clearer to say sthg like "Jetstream is enabled: xla test will fail".

**Review by dacorvo — APPROVED** [review]
Nice ! Thank you for this pull-request.

**Comment by tengomucho** [comment]
> I think I got lost in your changes: can you summarize how tests are now supposed to work ?

@dacorvo  as discussed offline, the idea is to change the default backend of TPU TGI from torch xla to jetstream. 
I just updated the tests so they use clearer markers to check if the backend they are running is correctly selected.

**Review by dacorvo — APPROVED** [review]

File: .github/workflows/test-pytorch-xla-tpu.yml [hunk 4]
```diff
@@ -31,4 +31,4 @@ jobs:
 
       - name: Build and test optimum tpu
         run: |
-          HF_TOKEN=${{ secrets.HF_TOKEN_OPTIMUM_TPU_CI }} make tests
+          JETSTREAM_PT_DISABLE=1 HF_TOKEN=${{ secrets.HF_TOKEN_OPTIMUM_TPU_CI }} make tests
```
Comment (baptistecolle) [inline .github/workflows/test-pytorch-xla-tpu.yml:34]: This should be on all test-pytorch-xla-* tests no?

https://github.com/huggingface/optimum-tpu/tree/main/.github/workflows

**Review by baptistecolle — COMMENTED** [review]
JETSTREAM_PT_DISABLE=1

File: text-generation-inference/tests/test_prefill_truncate.py [hunk 5]
```diff
@@ -1,35 +1,43 @@
...
+    _test_prefill_truncate()
+
+
+@skip_if_jetstream_pytorch_enabled
+def test_prefill_truncate():
+    _test_prefill_truncate()
```
[the remark this answers is not in the document]
Comment (tengomucho) [reply text-generation-inference/tests/test_prefill_truncate.py]: I do not understand the question. The test has been refactored after other reviews comments, I hope it's clearer now

**Review by baptistecolle — APPROVED** [review]
